File: mac/OCRReview/Services/PageStructureService.swift
import PDFKit

enum PageStructureService {
    static func remapAfterDelete(pages: [OCRPage], deletedPageNumber: Int) -> [OCRPage] {
        pages.compactMap { page in
            if page.pageNumber == deletedPageNumber { return nil }
            var updated = page
            if updated.pageNumber > deletedPageNumber {
                updated.pageNumber -= 1
            }
            return updated
        }
        .sorted { $0.pageNumber < $1.pageNumber }
    }

    static func remapAfterMove(pages: [OCRPage], fromIndex: Int, toIndex: Int, pageCount: Int) -> [OCRPage] {
        var slots: [OCRPage?] = (0..<pageCount).map { index in
            pages.first { $0.pageNumber == index + 1 }
        }
        guard fromIndex >= 0, fromIndex < slots.count, toIndex >= 0, toIndex < slots.count else {
            return pages
        }
        let item = slots.remove(at: fromIndex)
        slots.insert(item, at: toIndex)
        return slots.enumerated().compactMap { index, page in
            guard var updated = page else { return nil }
            updated.pageNumber = index + 1
            return updated
        }
    }
}

File: windows/src/OcrReview.Core/Services/PageStructureService.cs
using OcrReview.Core.Models;

namespace OcrReview.Core.Services;

public static class PageStructureService
{
    public static List<OcrPage> RemapAfterDelete(List<OcrPage> pages, int deletedPageNumber)
    {
        var result = new List<OcrPage>();
        foreach (var page in pages)
        {
            if (page.PageNumber == deletedPageNumber) continue;
            if (page.PageNumber > deletedPageNumber) page.PageNumber -= 1;
            result.Add(page);
        }
        return result.OrderBy(p => p.PageNumber).ToList();
    }

    public static List<OcrPage> RemapAfterMove(List<OcrPage> pages, int fromIndex, int toIndex, int pageCount)
    {
        if (fromIndex < 0 || fromIndex >= pageCount || toIndex < 0 || toIndex >= pageCount)
            return pages;

        var slots = new List<OcrPage?>(pageCount);
        for (int i = 0; i < pageCount; i++)
            slots.Add(pages.FirstOrDefault(p => p.PageNumber == i + 1));

        var item = slots[fromIndex];
        slots.RemoveAt(fromIndex);
        slots.Insert(toIndex, item);

        var result = new List<OcrPage>();
        for (int i = 0; i < slots.Count; i++)
        {
            var page = slots[i];
            if (page == null) continue;
            page.PageNumber = i + 1;
            result.Add(page);
        }
        return result;
    }
}
